lib/HTML/Restrict.pm

```diff
@@ -151,16 +151,38 @@ sub _build_parser {
 
                     foreach my $source_type ( 'href', 'src', 'cite' ) {
 
-                        if ( $attr->{$source_type} ) {
-                            my $uri = URI->new( $attr->{$source_type} );
-                            if ( defined $uri->scheme ) {
+                        my $link = $attr->{$source_type};
+
+                        # Remove unprintable ASCII control characters, which
+                        # are 1..8 and 14..31. These characters are not valid
+                        # in URLs, but they can prevent the URI parser from
+                        # recognizing the scheme when they are used as leading
+                        # characters.  Browsers will helpfully ignore them,
+                        # meaning that these characters can be used to defeat
+                        # HTML::Restrict when used as leading characters in a
+                        # link.  In our case we will strip them all regardless
+                        # of where they are in the URL. See
+                        # https://github.com/oalders/html-restrict/issues/30
+                        # https://url.spec.whatwg.org/
+                        # https://infra.spec.whatwg.org/#c0-control
+
+                        if ($link) {
+                            $link =~ s/[\00-\037]/ /g
+                                ;    # decimal 0..31 C0 control chars
+                            my $url = URI->new($link)->as_string;
+
+                            # The above regex doesn't strip the null byte
+                            $url =~ s{&#0;}{}g;
```

Comment:
I am not clear on why this is necessary. Also, it doesn't seem particularly effective given that `&#0000`, `&x0`, etc, are all equivalent. Maybe it isn't even worth worrying about if browsers don't seem to have the behavior on null bytes.

Reply:
Good point.  For some reason the regex above isn't catching it.  It doesn't seem to cause a problem for browsers, but the spec does say that it should be removed.